vpnglean/src/event.cpp

```diff
@@ -33,16 +26,45 @@ void EventMetric::record(const QJsonObject& extras) {
   // Helper vector to extend the lifetime of the strings
   // that hold the extra key values until they are used.
   QList<QByteArray> keepStringsAlive;
-  FfiExtra ffiExtras = m_parser->fromJsonObject(extras, keepStringsAlive);
+  FfiExtra ffiExtras(extras.count(), extras.count());
+
+  int count = 0;
+  foreach (const QString& key, extras.keys()) {
+    auto rawValue = extras.value(key);
+
+    if (rawValue.isString()) {
+      QByteArray value = rawValue.toString().toUtf8();
+      keepStringsAlive.append(value);
+      ffiExtras.values[count] = value.constData();
+    } else if (rawValue.isBool()) {
+      ffiExtras.values[count] = rawValue.toBool() ? "true" : "false";
+    } else if (rawValue.isDouble()) {
+      QByteArray value = QString::number(rawValue.toDouble()).toUtf8();
+      keepStringsAlive.append(value);
+      ffiExtras.values[count] = value.constData();
+    } else {
+      Q_ASSERT(false);
+      // TODO: Record error.
+      continue;
+    }
+
+    QByteArray k = key.toUtf8();
+    keepStringsAlive.append(k);
+    ffiExtras.keys[count] = k.constData();
 
+    count++;
+  }
+
+  ffiExtras.values.resize(count);
+  ffiExtras.keys.resize(count);
   if (!ffiExtras.keys.empty()) {
-#if not(defined(MZ_WASM) || defined(BUILD_QMAKE))
+#if not(defined(__wasm__) || defined(BUILD_QMAKE))
     glean_event_record(m_id, ffiExtras.keys.data(), ffiExtras.values.data(),
                        ffiExtras.keys.size());
 #endif
   } else {
-    logger.error() << "Attempted to record an event with extras, but no extras "
-                      "were provided. Ignoring.";
+    qWarning() << "Attempted to record an event with extras, but no extras "
```

Comment:
Why switch from the logger?

Reply:
The general idea here is that `vpnglean` should be a standalone library, which means that it can't depend on anything in the VPN app or we risk creating an unsolvable dependency graph.

Reply:
Got it, hm. It would be overkill to turn the logger into a santdalone library as well so that we can use it here... However, would you mind adding a quick comment as to why we are going with a different logging method just here?